crates/red_knot_python_semantic/src/semantic_index/definition.rs

```diff
@@ -1,57 +1,133 @@
-use crate::semantic_index::ast_ids::{
-    ScopedClassId, ScopedExpressionId, ScopedFunctionId, ScopedStatementId,
-};
+use ruff_db::parsed::ParsedModule;
+use ruff_db::vfs::VfsFile;
+use ruff_python_ast as ast;
 
-#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
-pub enum Definition {
-    Import(ImportDefinition),
-    ImportFrom(ImportFromDefinition),
-    ClassDef(ScopedClassId),
-    FunctionDef(ScopedFunctionId),
-    Assignment(ScopedStatementId),
-    AnnotatedAssignment(ScopedStatementId),
-    NamedExpr(ScopedExpressionId),
-    /// represents the implicit initial definition of every name as "unbound"
-    Unbound,
-    // TODO with statements, except handlers, function args...
+use crate::ast_node_ref::AstNodeRef;
+use crate::node_key::NodeKey;
+use crate::semantic_index::symbol::{FileScopeId, ScopedSymbolId};
+
+#[salsa::tracked]
+pub struct Definition<'db> {
+    /// The file in which the definition is defined.
+    #[id]
+    pub(super) file: VfsFile,
+
+    /// The scope in which the definition is defined.
+    #[id]
+    pub(crate) scope: FileScopeId,
+
+    /// The id of the corresponding symbol. Mainly used as ID.
+    #[id]
+    symbol_id: ScopedSymbolId,
+
+    #[no_eq]
```

Comment:
What are the implications of using `no_eq` here?

I'm not totally clear on how this all works, but it's important that two different definitions of the same symbol in the same scope are still considered different definitions.

Reply:
The `no_eq` doesn't influence the definition's identity.

You can think of each field of a salsa-tracked as a salsa query. A query depending on the `symbol_id` field only gets re-executed if the `symbol_id` at `t1` is not equal to the `symbol_id` at `t2`. 

The `no_eq` removes that optimization, and any query depending on the `Definition` will re-execute whenever the definition changes. Which is whenever the AST changes and that seems correct.